dbsc/workflow/rules/mrtpipelines.smk

```diff
@@ -1,571 +1,600 @@
+            datatype="dwi",
+            suffix="brainmask.mif",
+            **config["subj_wildcards"]
+        ),
+    threads: workflow.cores
+    group:
+        "subject_1"
     container:
-        config['singularity']['mrtpipelines']
-    shell: 
-        'mrconvert -nthreads {threads} -fslgrad {input.bvecs} {input.bvals} {input.dwi} {output.dwi} '
-        'mrconvert -nthreads {threads} {input.mask} {output.mask}'
+        config["singularity"]["mrtrix"]
+    shell:
+        "mrconvert -nthreads {threads} -fslgrad {input.bvecs} {input.bvals} {input.dwi} {output.dwi} && "
+        "mrconvert -nthreads {threads} {input.mask} {output.mask}"
+
 
+rule dwi2response:
+    """
+    Estimate response functions using Dhollander algorithm
 
-rule estimate_response:
-    """Estimate response functions using Dhollander algorithm """
+    Dhollander, T.; Mito, R.; Raffelt, D. & Connelly, A. Improved white matter response function estimation for 3-tissue constrained spherical deconvolution. Proc Intl Soc Mag Reson Med, 2019, 555
+    """
     input:
-        dwi=rules.nii_to_mif.output.dwi,
-        mask=rules.nii_to_mif.output.mask
+        dwi=rules.nii2mif.output.dwi,
+        mask=rules.nii2mif.output.mask,
     params:
-        threads=workflow.cores,
+        shells=shells,
+        lmax=lmaxes,
     output:
-        sfwm=bids(
-            root=join(config['output_dir'], 'mrtpipelines'),
-            datatype='dwi',
-            desc='sfwm',
-            suffix='response.txt',
-            **config['subj_wildcards'],
-        )
-        gm=bids(
-            root=join(config['output_dir'], 'mrtpipelines'),
-            datatype='dwi',
-            desc='gm'
-            suffix='response.txt',
-            **config['subj_wildcards'],
-        )
-        csf=bids(
-            root=join(config['output_dir'], 'mrtpipelines'),
-            datatype='dwi',
-            desc='csf'
-            suffix='response.txt',
-            **config['subj_wildcards'],
-        )
-    group: "subject_1"
+        wm_rf=bids(
+            root=mrtrix_dir,
+            datatype="response",
+            desc="wm",
+            suffix="response.txt",
+            **config["subj_wildcards"],
+        ),
+        gm_rf=bids(
+            root=mrtrix_dir,
+            datatype="response",
+            desc="gm",
+            suffix="response.txt",
+            **config["subj_wildcards"],
+        ),
+        csf_rf=bids(
+            root=mrtrix_dir,
+            datatype="response",
+            desc="csf",
+            suffix="response.txt",
+            **config["subj_wildcards"],
+        ),
+    threads: workflow.cores
+    group:
+        "subject_1"
     container:
-        config['singularity']['mrtpipelines']
+        config["singularity"]["mrtrix"]
     shell:
-        'dwi2response dhollander {input.dwi} {output.sfwm} {output.gm} {output.csf} -nthreads {threads}'
+        'if [ {params.shell} == "" ] && [ {params.lmax} == ""]; then'
```

Comment:
I think this would be a bit cleaner if you were to get rid of the if statement and provide a sensible default for `shell` and `lmax` where they're defined on line 13.

Reply:
I agree that this isn't the cleanest. That said, I would rather avoid providing a default within the workflow as Mrtrix itself has a mechanism in their code to provide a default that works with the input diffusion data (e.g. dealing with 2 shells vs 3 shells without explicitly defining the number of shells).

Reply:
Gotcha. In that case, could do:
```
params:
    shells=f"-shells {shells}" if shells else ""
    lmax=f"-lmax {lmax}" if lmax else ""
```

Reply:
Would that not pass "" to the command? I was running into that issue before and that ended up causing issues.

Reply:
I would change the shell directive to:
`"dwi2response dhollander {input.dwi} {output.wm_rf} {output.gm_rf} {output.csf_rf} -nthreads {threads} -mask {input.mask} {params.shells} {params.lmax} "`,
which I would expect to resolve to:
`"dwi2response dhollander {input.dwi} {output.wm_rf} {output.gm_rf} {output.csf_rf} -nthreads {threads} -mask {input.mask}  "`
if `params.shells` and `params.lmax` were both an empty string, but it's possible I'm incorrect.

Reply:
Well, we will find out when we test this.